core/src/event/callback_registry.rs

```diff
@@ -210,3 +226,130 @@ impl EventCallbackRegistry {
         self.complete()
     }
 }
+
+// --------------------------------
+// "Native" Trace Callback Registry
+// --------------------------------
+
+#[derive(Debug, Clone)]
+pub struct TraceResult {
+    pub from: Address,
+    pub to: Address,
+    pub value: U256,
+    pub tx_information: TxInformation,
+    pub found_in_request: LogFoundInRequest,
+}
+
+impl TraceResult {
+    /// Create a "NativeTokenTransfer" TraceResult for sinking and streaming.
+    pub fn new_native_transfer(
+        action: &Call,
+        trace: &Trace,
+        network: &str,
+        start_block: U64,
+        end_block: U64,
+    ) -> Self {
+        Self {
+            from: action.from,
+            to: action.to,
+            value: action.value,
+            tx_information: TxInformation {
+                network: network.to_string(),
+                address: Address::zero(),
+                block_number: U64::from(trace.block_number),
+                block_timestamp: None,
+                transaction_hash: trace.transaction_hash.expect("checked prior"),
+                block_hash: trace.block_hash,
+                transaction_index: U64::from(trace.transaction_position.unwrap_or(0)),
+                log_index: U256::from(0),
+            },
+            found_in_request: LogFoundInRequest { from_block: start_block, to_block: end_block },
+        }
+    }
+}
+
+pub type TraceCallbackResult<T> = Result<T, String>;
+
+#[derive(Clone)]
+pub struct TraceCallbackRegistryInformation {
+    pub id: String,
+    pub indexer_name: String,
+    pub event_name: String,
+    pub contract_name: String,
+    pub trace_information: TraceInformation,
+    pub callback: TraceCallbackType,
+}
+
+impl TraceCallbackRegistryInformation {
+    pub fn info_log_name(&self) -> String {
+        format!("{}::{}", self.indexer_name, self.event_name)
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct TraceCallbackRegistry {
+    pub events: Vec<TraceCallbackRegistryInformation>,
+}
+
+impl TraceCallbackRegistry {
```

Comment:
seems a direct copy and paste of EventCallbackRegistry maybe we can share this a bit so we dont have 2 places to manage the code?

Reply:
I'll extract it out to a function to share the logic, part of the challenge I had here (as I initially wanted to use 100% shared code for this) is backwards compatibility.

If I try to completely unify the `CallbackRegistry` and modify the type / input types to the `callback` itself, it will become a breaking change for any Rust bindings project types (nocode projects should be fine).

I opted to keep it non-breaking, so am limited in the refactor, but can definitely extract the duplicate code to a shared fn 👍

Reply:
i think if they just generated the new typings it would fix itself as its not a breaking change in the interface level right? aka rindexer codegen typings